### PR TITLE
feat: Migration guide for Foundry projects

### PR DESCRIPTION
**Pull request by AnastasiiaVashchuk**
What:
 - A migration guide for Foundry projects was added. It is short since
1)  all the basic info is present on the getting started page 
2) The discrepancies guide is not well-formed yet.
- Existing guides have been updated with up-to-date info.
- added some notes(note about foundry book, support section etc)

### CODE REVIEW

File: content/00.build/40.tooling/30.foundry/10.overview.md [hunk 1]
```diff
@@ -15,41 +19,83 @@ deployment, testing, and interaction with smart contracts on ZKsync.
 It is open-sourced, and contributions from the developer community are welcome.
 For more details and contributions, visit the [GitHub repository](%%zk_git_repo_foundry-zksync%%).
 
-## Features and Limitations
+### Changes Made
+
+To use for ZKsync environments, include `--zksync` when running `forge` or `vm.zkVm(true)` in tests. The modifications include:
```
[the remark this answers is not in the document]
Comment (AnastasiiaVashchuk) [reply content/00.build/40.tooling/30.foundry/10.overview.md:24]: When was this? We use `ZKsync ` for branding purposes

[the remark this answers is not in the document]
Comment (nbaztec) [reply content/00.build/40.tooling/30.foundry/10.overview.md:24]: I believe that was early March, where we wanted to select a single way to write zksync and zkVM/zkEVM. I don't have any input for its correctness, but would also be nice for us to know the recommended way to brand "zksync"

File: content/00.build/40.tooling/30.foundry/10.overview.md [hunk 2]
```diff
@@ -15,41 +19,83 @@ deployment, testing, and interaction with smart contracts on ZKsync.
...
+
+1. **Compilation:** `solc` and `zksolc` are used for compiling. The resulting bytecodes are combined into `DualCompiledContract`
+and managed through `Executor` to `CheatcodeTracer`.
+2. **EVM Interactions:**
+   - EVM calls are standard except for `address.balance` and `block.timestamp`/`block.number`, which pull data from ZKsync
```
[the remark this answers is not in the document]
Comment (AnastasiiaVashchuk) [reply content/00.build/40.tooling/30.foundry/10.overview.md:29]: @nbaztec, do you have an idea of the best approach here?

[the remark this answers is not in the document]
Comment (AnastasiiaVashchuk) [reply content/00.build/40.tooling/30.foundry/10.overview.md:29]: I took this info from the repo README. Who is this information for?

File: content/00.build/40.tooling/30.foundry/10.overview.md [hunk 3]
```diff
@@ -15,41 +19,83 @@ deployment, testing, and interaction with smart contracts on ZKsync.
...
+(ZK-storage and ZK-specific context, respectively).
+3. **Transaction Handling:**
+   - `CALL` and `CREATE` operations are captured and converted to ZKsync transactions. This process includes fetching
+ZKsync-equivalent bytecode, managing account nonces, and marking EOA appropriately to comply with ZKsync requirements.
+4. **Execution and State Management:**
+   - ZKsync VM processes the transaction and returns state changes, which are applied to `journaled_state`. Results are relayed back.
```
[the remark this answers is not in the document]
Comment (AnastasiiaVashchuk) [reply content/00.build/40.tooling/30.foundry/10.overview.md:35]: I took this info from the repo README. Who is this information for?

File: content/00.build/40.tooling/30.foundry/10.overview.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (nbaztec) [reply content/00.build/40.tooling/30.foundry/10.overview.md:29]: I'd stay away from naming the exact constructs like `DualCompiledContracts`, `Executor`, `CheatcodeTracer` etc. and focus on macro-functionality from an end user's perspective, i.e.:

* `solc` and `zksolc` both are used to compile the solidity files to provide interoperatability between EVM which foundry starts with and the switchover to zkEVM when calls are performed within the test. EVM opcode calls like `address.balance`, `block.timestamp`/`block.number` are translated seamlessly to their zksync counterparts when run within zksync mode.

...as a general idea we should stick to familiar concepts which users intuitively understand, and how they apply to foundry-zksync (if required due to changes that need to be made to adpat an EVM test to zkEVM)

File: content/00.build/40.tooling/30.foundry/10.overview.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (nbaztec) [reply content/00.build/40.tooling/30.foundry/10.overview.md:35]: This was originally meant to document the architecture choices and decisions, when trying to communicate with the foundry team, in order to merge back the changes

File: content/00.build/40.tooling/30.foundry/30.discrepancies.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+We explicitly avoid using `prop_filter` as it can cause the same "too many rejections" failure in `proptest`,
+and instead opt to map the invalid addresses in the valid address range, so the fuzzer's *performance is practically unaffected*.
+
+A side-effect of this is that `address(0)` would no longer be the boundary value. However, since we support foundry's
+[inline test config](https://book.getfoundry.sh/reference/config/inline-test-config) the option can be selectively applied to certain tests as required.
+
+## Contract Size limitations
+In ZkEVM, the contract size must be divisible by 32 bytes (4 instructions). For example,  `bytes("mock code")` would fail the `ensure_chunkable` validation.
+Additionally, the bytecode cannot exceed `2^16` 32-byte words.
```
Comment (nbaztec) [inline content/00.build/40.tooling/30.foundry/30.discrepancies.md:59]: This might need to be clarified with the zksolc compiler team. From what I'm aware of, there's no direct metric that can tell a user if their contract is too big. The only way to figure this out is to perform the compilation and observe the `bytecode cannot exceed 65536` error (if any).

File: content/00.build/60.test-and-debug/50.foundry.md [hunk 5]
```diff
@@ -17,9 +17,9 @@ For more detailed documentation related to Foundry testing please refer to the o
 ---
 ## Cheatcodes
 
-Cheatcodes allow you to change the block number, your identity, and more.
-`foundry-zksync` supports the most common Foundry cheatcodes.
-For an exhaustive list of supported cheatcodes refer to the [Supported Cheatcodes for foundry-zksync](%%zk_git_repo_foundry-zksync%%/blob/main/SUPPORTED_CHEATCODES.md).
+We support cheatcodes on the test code (because it is run on the EVM).
+But we do not support them on contracts that the test creates/invokes(because they run on the zkVM).
```
[the remark this answers is not in the document]
Comment (nbaztec) [reply content/00.build/60.test-and-debug/50.foundry.md:21]: Another point where I've seen zkEVM but we were using zkVM before. Could use clarification.

File: content/00.build/40.tooling/30.foundry/20.getting-started.md [hunk 6]
```diff
@@ -229,7 +151,7 @@ contract Factory {
 **Compile `GreeterFactory.sol`:**
 
 ```bash
-forge build --is-system=true --zksync
+forge build --system-mode=true --zksync
```
[the remark this answers is not in the document]
Comment (dutterbutter) [reply content/00.build/40.tooling/30.foundry/20.getting-started.md:154]: Ahh yes it's been changed to enable-eravm-extension. 

Will update.

**Review by itsacoyote — CHANGES_REQUESTED** [review]
Various little edits here and there. Also, for writing titles and headings, we want to use sentence casing - https://developers.google.com/style/capitalization#capitalization-in-titles-and-headings

I know we don't have it fixed across the site yet so it's inconsistent, but would like to ensure we do this when we come across or add new content in PRs.

File: content/00.build/40.tooling/30.foundry/40.migration-guide/30.testing.md [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+
+2. **Fuzz Configuration**:
+   Alternatively, you can set the `no_zksync_reserved_addresses` option in your fuzz configuration to automatically exclude addresses in the reserved range.
+
+   ```toml
+   [profile.default]
+   fuzz = { no_zksync_reserved_addresses = true }
+   ```
+
+These approaches ensure that fuzz testing avoids the reserved kernel space, preventing test failures and ensuring compliance with ZKsync's address rules.
```
Comment (itsacoyote) [inline content/00.build/40.tooling/30.foundry/40.migration-guide/30.testing.md:388]: There's a red highlight here in github, I'm guessing it's because there's no newline at the end of this file?

**Review by itsacoyote — CHANGES_REQUESTED** [review]

File: content/00.build/40.tooling/30.foundry/40.migration-guide/30.testing.md (hunk 7 above)
Comment (dutterbutter) [reply content/00.build/40.tooling/30.foundry/40.migration-guide/30.testing.md:388]: hmm there is a newline at the end of the file? I don't see any issue in my editor either.

File: content/00.build/40.tooling/30.foundry/40.migration-guide/30.testing.md [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+   Alternatively, you can set the `no_zksync_reserved_addresses` option in your fuzz configuration to automatically exclude addresses in the reserved range.
+
+   ```toml
+   [profile.default]
+   fuzz = { no_zksync_reserved_addresses = true }
+   ```
+
+These approaches ensure that fuzz testing
+avoids the reserved kernel space,
+preventing test failures and ensuring compliance with ZKsync's address rules.
```
Comment (dutterbutter) [inline content/00.build/40.tooling/30.foundry/40.migration-guide/30.testing.md:390]: @itsacoyote weird. I just pushed a commit ensuring there was a new line but its not showing up here? 

Screenshot 2024-09-18 at 9 08 09 AM

Comment (dutterbutter) [reply content/00.build/40.tooling/30.foundry/40.migration-guide/30.testing.md:390]: It seems to pass the CI build and linting tho?

Comment (itsacoyote) [reply content/00.build/40.tooling/30.foundry/40.migration-guide/30.testing.md:390]: 🤷🏻 if it's not breaking any of our CI and linting, let's leave it then 👍🏻

**Review by itsacoyote — APPROVED** [review]
🎉
